Add support for 1-second Storage Resolution in the AWS EMF Exporter

jpbarto opened the pull request:
This change implements the capability for users of the AWS EMF Exporter to specify which metrics they would like to have sent to CloudWatch with a 1 second Storage Resolution. The EMF Exporter now explicitly states the Storage Resolution for each metric as 60 seconds, the previous implicit default, so there is no behavior change. If the user specifies a metric to have 1 second resolution it will be sent to CloudWatch EMF with the Storage Resolution set accordingly.

#### Description
Previously the AWS EMF Exporter sent metric data into CloudWatch without specifying the storage resolution. CloudWatch would then default to a 60 second storage resolution, even if metrics are sent more frequently than every 60 seconds. This would confuse users when they try to apply functions like AVG, SUM, MAX, or MIN to their metrics with a period of 5 seconds. The function would be applied by CloudWatch to 60 seconds worth of data and produced unexpected results and confusion for the user. This commit makes this 60 second resolution explicit in the messages sent to CloudWatch by the EMF Exporter and also gives the user the option to specify a more granular 1 second resolution per metric in the configuration file of the AWS EMF Exporter.

#### Link to tracking issue
Fixes #29506 

#### Testing
Added tests to verify that config file parsing validates a metric descriptor that specifies either a valid unit, valid storage resolution, or both and rejects other invalid metric descriptors.
Added tests that the translation from metric data to CW EMF carries a storage resolution with it, defaulting to a value of 60 (current behavior) if no storage resolution valid is explicitly set in the configuration.

#### Documentation
Comments added in the code but have not updated the README.md pending acceptance of the PR.

jpbarto committed:
Add support for AWS EMF 1 second Storage Resolution in the AWS EMF Exporter

This change implements the capability for users of the AWS EMF Exporter to specify
which metrics they would like to have sent to CloudWatch with a 1 second Storage
Resolution. The EMF Exporter now explicitly states the Storage Resolution for each
metric as 60 seconds, the previous implicit default, so there is no behavior
change. If the user specifies a metric to have 1 second resolution it will be
sent to CloudWatch EMF with the Storage Resolution set accordingly.

Previously the AWS EMF Exporter sent metric data into CloudWatch without
specifying the storage resolution. CloudWatch would then default to a 60 second
storage resolution, even if metrics are sent more frequently than every 60
seconds. This would confuse users when they try to apply functions like AVG,
SUM, MAX, or MIN to their metrics with a period of 5 seconds. The function would
be applied by CloudWatch to 60 seconds worth of data and produced unexpected
results and confusion for the user. This commit makes this 60 second resolution
explicit in the messages sent to CloudWatch by the EMF Exporter and also gives
the user the option to specify a more granular 1 second resolution per metric
in the configuration file of the AWS EMF Exporter.

Aneurysm9 commented on exporter/awsemfexporter/config.go:
```diff
@@ -101,6 +101,8 @@ type MetricDescriptor struct {
 	// Overwrite set to true means the existing metric descriptor will be overwritten or a new metric descriptor will be created; false means
 	// the descriptor will only be configured if empty.
 	Overwrite bool `mapstructure:"overwrite"`
+	// Set the storage resolution for this metric in AWS Embedded Metric Format (EMF). Valid values are 1 (for 1 second resolution) and 60 (for 60 second resolution).
+	StorageResolution int `mapstructure:"storage_resolution"`
```
The `MetricDescriptor` configuration is extremely limited and can be accomplished through generally-applicable processors.  I think it should be deprecated and slated for removal at some point in the future so we should not expand its surface area any further.

jpbarto replied:
The latest update instead looks for a metric attribute named `aws.emf.storage_resolution` which can be applied either in the application's code or using the Attribute Processor in the collector itself.

Aneurysm9 commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -48,7 +48,7 @@ type cWMetrics struct {
 	Namespace  string
 	Dimensions [][]string
-	Metrics    []map[string]string
+	Metrics    []map[string]any
```
I don't like the use of `any` here.  Use the type system to prevent invalid data from being represented.  If that means a new type needs to be created then a new type will need to be created.  Rather than `[]map[string]any` this should be something like `[]metricDefinition` with `metricDefinition` defined like this:

```go
type metricDefinition struct {
  Name string
  Unit string
  StorageResolution int
}
```

jpbarto replied:
The latest update creates a `cWMetricInfo` struct in the translator to mimic the `metricInfo` struct in grouped_metric.go, replacing the use of `map[string]any` in the translator.

jpbarto committed:
Add EMF Storage Resolution support using Metric Attributes

This commit addresses comments in PR#36057 where it was suggested that metric description for the EMF Exporter
should be deprecated and metric attributes used to specify the storage resolution for each metric.

The code has been modified to remove the original changes and instead have the exporter look for a metric
attribute named `aws.emf.storage_resolution` and use its value for the storage resolution when sending
a message to CloudWatch EMF.

The code has also added a new struct `cWMetricInfo` to hold the metric name, unit, and storage resolution
where previously this was a `[]map[string]string` type.

removed unnecessary changes

jpbarto committed:
Merge branch 'main' into awsemf_storageresolution

jpbarto commented:
@Aneurysm9 are you able to approve the change in its current form which now uses metric attributes to define storage resolution?

jpbarto committed:
Merge branch 'main' into awsemf_storageresolution

mxiamxia commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -208,8 +221,13 @@ func groupedMetricToCWMeasurement(groupedMetric *groupedMetric, config *Config)
 		dimSet[idx] = labelName
 		idx++
 	}
+
+	fmt.Printf("##--## Dim Set: %+v\n", dimSet)
+
```
Please remove all the debug logs

mxiamxia commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -467,3 +499,15 @@ func translateGroupedMetricToEmf(groupedMetric *groupedMetric, config *Config, d
+func filterAWSEMFAttributes(labels map[string]string) map[string]string {
+	// remove any labels that are attributes specific to AWS EMF Exporter
+	filteredLabels := make(map[string]string)
+	emfAttributeFilterRE := regexp.MustCompile(emfAttributeFilter)
+	for labelName := range labels {
+		if !emfAttributeFilterRE.MatchString(labelName) {
+			filteredLabels[labelName] = labels[labelName]
+		}
+	}
```
I'm concerned about the performance impact of applying regex matching on every metric attribute for each grouped metric data point. 
Is there any requirement to add more `aws.emf.*` attribute patterns? Could you perform an exact match for `labelname == "aws.emf.storage_resolution"` and break the loop once it's found?

mxiamxia commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -208,8 +221,13 @@ func groupedMetricToCWMeasurement(groupedMetric *groupedMetric, config *Config)
+
 	dimensions := [][]string{dimSet}
 
+	fmt.Printf("##--## Dimensions: %+v", dimensions)
```
ditto

mxiamxia commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -228,14 +246,22 @@ func groupedMetricToCWMeasurement(groupedMetric *groupedMetric, config *Config)
 	dimensions = append(dimensions, rollupDimensionArray...)
 
-	metrics := make([]map[string]string, len(groupedMetric.metrics))
+	fmt.Printf("##--## Post Dimensions: %+v", dimensions)
```
ditto

mxiamxia commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -29,6 +31,11 @@ const (
+
+	// metric attributes for AWS EMF, not to be treated as metric labels
+	// emfAttributeFilterRE          = regexp.MustCompile("^awsemf.[A-Za-z_]*$")
+	emfAttributeFilter            = "^awsemf.[A-Za-z_]*$"
```
q: why is regex rule is `^awsemf.[A-Za-z_]*$` instead of `^aws.emf.[A-Za-z_]*$`?

jpbarto replied:
Thank you @mxiamxia - I've swapped the more complex regex for a simpler comparison operator. As new metric attributes are added this can be expanded to a switch statement which will likely still perform better than a compiled regex.

jpbarto committed:
Replaced the use of regex to match EMF metric attributes for a more performant comparison operator

jpbarto committed:
Merge remote-tracking branch 'refs/remotes/origin/awsemf_storageresolution' into awsemf_storageresolution

mxiamxia reviewed (approved):
LGTM. ty!

jpbarto committed:
Merge branch 'main' into awsemf_storageresolution

jpbarto commented:
@andrzej-stencel good morning, are you able to review this pull request please?

andrzej-stencel commented:
@jpbarto this probably deserves a changelog entry? Can you add one please https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#changelog.

jpbarto committed:
Merge branch 'open-telemetry:main' into awsemf_storageresolution

jpbarto committed:
updated changelog to reflect new storage resolution feature added to aws emf exporter

jpbarto commented:
@andrzej-stencel the changelog has been updated as requested.

jpbarto commented:
@andrzej-stencel @mxiamxia the PR requires a second reviewer to approve the pull request. This is currently assigned to @bryan-aguilar. Bryan was involved with the original issue raised last year and has been idle on GitHub since June of this year. Can I please ask that this PR be assigned to an alternate reviewer?

jpbarto committed:
repaired partially merged changelog entry for storage resolution feature of emf exporter

jpbarto commented:
@andrzej-stencel this PR has been approved by both yourself and one of the code owners, Min Xia (see above). Please could you mark this PR as `ready to merge`?

jpbarto committed:
Merge branch 'main' into awsemf_storageresolution
